### PR TITLE
Wait for RC to stabilize in rolling-update e2e

### PR DESCRIPTION
**Pull request by bprashanth**
https://github.com/kubernetes/kubernetes/issues/25170

### CODE REVIEW

File: test/e2e/framework/util.go [hunk 1]
```diff
@@ -1005,6 +1005,21 @@ func waitForRCPodOnNode(c *client.Client, ns, rcName, node string) (*api.Pod, er
 	return p, err
 }
 
+// WaitForRCToStabilize waits till the RC has a matching generation/replica count between spec and status.
+func WaitForRCToStabilize(c *client.Client, ns, name string, interval, timeout time.Duration) error {
+	return wait.PollImmediate(interval, timeout, func() (bool, error) {
```
Comment (lavalamp) [inline test/e2e/framework/util.go]: Use a watch instead? Search for watch.Until in this file for examples

**Comment by lavalamp** [comment]
I haven't looked super carefully yet, but I think we can troll people by just continuously scaling an RC up and down, causing it to never increment the observedGeneration?

Another way for a client to fix this is to not do the second action until the first action has stabilized. This doesn't work if multiple clients make simultaneous changes.

Something that I think *does* work is to publish the count of pending actions.

**Comment by bprashanth** [comment]
> I haven't looked super carefully yet, but I think we can troll people by just continuously scaling an RC up and down, causing it to never increment the observedGeneration?

Anytime it goes to create/delete another batch of pods it will update generation, just not when it's created some and waiting to hear about them. So it will create/delete, as each create happens update status.Replicas, once all creates are done (so expectations are 0)- update generation, start processing next batch of create/delete ...it's certainly possible to confuse the client.

> Another way for a client to fix this is to not do the second action until the first action has stabilized. This doesn't work if multiple clients make simultaneous changes.

Hence the accompanying e2e change. But that feels like a cop out.

> Something that I think does work is to publish the count of pending actions

Yeah I guess that would work too, but updating generation this way is strictly better anyway?

**Comment by lavalamp** [comment]
> So it will create/delete, as each create happens update status.Replicas, once all creates are done (so expectations are 0)- update generation, start processing next batch of create/delete ...it's certainly possible to confuse the client.

It seems like this still allows for a bunch of update sequences that could confuse a client. Maybe up/down doesn't break clients any more, but up/up/down might still?

Additionally, with this change I'm no longer sure when observedGeneration is set. It used to be set "when the controller observed the corresponding spec" but not necessarily when the controller had acted upon it. Now you're changing that, but I'm not sure how to put the new rule concisely.

I think listing ObservedGeneration as "the lowest generation for which actions are still pending" would be best, but I think that requires tracking adds/deletes individually rather than in bulk.

>> Something that I think does work is to publish the count of pending actions

> Yeah I guess that would work too, but updating generation this way is strictly better anyway?

No, adding "pending action count == 0" as a condition is much clearer to a client, then no sequence of resizings will fool a client. The drawback there is that it might never get to zero if the background pod churn rate is sufficiently high.

**Comment by bprashanth** [comment]
Yeah I agree clients can get foxed. I was going for: best effort at consistent view + best practices before deleting an rc. Maybe we should keep this to just the e2e fix and discuss on the issue? We could always change this to mean: genration is updated only when a given spec is *fulfilled*. That would involve more book keeping though.

**Comment by bprashanth** [comment]
Also I think the predominant use case will be the one client that does scale up/down (kubectl rolling-update)

**Comment by bprashanth** [comment]
Alright, this is now just the e2e fix to wait till the rc has stabilized. PTAL.

**Comment by lavalamp** [comment]
"fulfilled" is a tricky concept, since the natural churn rate may be > 1
pod at basically all times for large enough RCs.

**Comment by lavalamp** [comment]
But yes I think just fixing the e2e and discussing a general solution
further in the issue is a good idea.

**Comment by bprashanth** [comment]
I removed the generation number changes, tests are green

**Comment by lavalamp** [comment]
LGTM thanks

**Comment by bprashanth** [comment]
I guess test flake fixes get prio

**Comment by wojtek-t** [comment]
It seems that unit tests didn't started (at least I'm not able to get their logs).

@k8s-bot unit test this please github issue: #IGNORE

**Comment by bprashanth** [comment]
Last failure has:
```
ERROR: (gcloud.compute.instances.create) Some requests did not succeed:
 - Quota 'IN_USE_ADDRESSES' exceeded.  Limit: 150.0
```
So that looks like https://github.com/kubernetes/kubernetes/issues/25629 again

**Comment by pwittrock** [comment]
@k8s-bot unit test this issue: #25363

[artifacts](https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/25171/kubernetes-pull-test-unit-integration/27773/)

**Pull request merged**